rs/bitcoin/kyt/src/state.rs

```diff
@@ -52,37 +52,107 @@ pub struct FetchedTx {
 // Max number of concurrent http outcalls.
 const MAX_CONCURRENT: u32 = 50;
 
+// Default expiry of cached fetch transaction entries is 7 days (in nano seconds).
+const DEFAULT_FETCH_TX_EXPIRY: u64 = 7 * 24 * 3600 * 1000000;
+
+// Max number of entries in the cache is set to 10_000. Since the average transaction size
+// is about 400 bytes, the estimated memory usage of the cache is in the order of 10s of MBs.
+const MAX_FETCH_TX_ENTRIES: usize = 10_000;
+
 // The internal KYT state includes:
 // 1. Outcall capacity, a semaphore limiting max concurrent outcalls.
 // 2. fetch transaction status, indexed by transaction id.
 //
 // TODO(XC-191): persist canister state
 thread_local! {
     static OUTCALL_CAPACITY: RefCell<u32> = const { RefCell::new(MAX_CONCURRENT) };
-    static FETCH_TX_STATUS: RefCell<BTreeMap<Txid, FetchTxStatus>> = RefCell::new(BTreeMap::default());
+    static FETCH_TX_CACHE: RefCell<FetchTxCache<FetchTxStatus>> = RefCell::new(
+        FetchTxCache::new(DEFAULT_FETCH_TX_EXPIRY, MAX_FETCH_TX_ENTRIES)
+    );
+}
+
+// Time in nanoseconds since the epoch (1970-01-01).
+type Timestamp = u64;
+
+// Time duration in nanoseconds.
+type Duration = u64;
+
+struct FetchTxCache<T> {
+    expiry: Duration,
+    max_entries: usize,
+    status: BTreeMap<Txid, T>,
+    created: VecDeque<(Txid, Timestamp)>,
```

Comment:
Why not simply an LRU cache?
The only advantage of expiring entries that I can see is that we reduce the memory footprint and save a few cycles. Is that the motivation or am I missing something?

Reply:
Good question! I initially only had `expiry` but later I added `max_entries`.

I had two objectives:
1. when a caller has to call check_transaction again, its previous (already paid) effort would still be in the cache.
2. the max bound should be straightforward to reason about.

A reasonable expiry ensures 1, and a reasonable max_entries ensures 2. But looking at it again, a sufficiently large `max_entries` seems enough for 1 too.

Reply:
Right now, an entry is purged when either condition satisfies (oldest entry expired or we reach max_entries). But thinking again, maybe it should be when both conditions satisfy?

Reply:
No, certainly not both conditions. That would mean that the cache could theoretically grow to an unbounded size.
The maximum size is the main factor. We can also use an age constraint but that is a secondary concern. Do you agree?

Reply:
Yes, that make sense. I'll just go for `max_entries` (essentially LRU) to make it simple.

Reply:
Changed to a simple purge when cache is at capacity.